source/material_model/steinberger.cc

```diff
@@ -501,21 +500,41 @@ namespace aspect
       double cp = 0.0;
       if (!latent_heat)
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->specific_heat(temperature,pressure);
+          if (material_file_names.size() == 1)
+            {
+              cp = material_lookup[0]->specific_heat(temperature,pressure);
+            }
+          else if (material_file_names.size() == compositional_fields.size() + 1)
+            {
+              const double background_cp = material_lookup[0]->specific_heat(temperature,pressure);
+              cp = background_cp;
+              for (unsigned int i = 0; i < compositional_fields.size(); ++i)
+                cp += compositional_fields[i] *
+                      (material_lookup[i+1]->specific_heat(temperature,pressure) - background_cp);
+            }
           else
             {
-              for (unsigned i = 0; i < n_material_data; i++)
+              for (unsigned i = 0; i < material_file_names.size(); i++)
                 cp += compositional_fields[i] * material_lookup[i]->specific_heat(temperature,pressure);
             }
         }
       else
         {
-          if (n_material_data == 1)
-            cp = material_lookup[0]->dHdT(temperature,pressure);
+          if (material_file_names.size() == 1)
```

Comment:
Well, I could create the following:
```
if ((material_lookup.size() == 1) ||
    (material_lookup.size() == compositional_fields.size() + 1))
  {
            const double background_cp = material_lookup[0]->specific_heat(temperature,pressure);
            cp = background_cp;
            if (material_lookup.size() == compositional_fields.size() + 1)
              for (unsigned int i = 0; i < compositional_fields.size(); ++i)
                cp += compositional_fields[i] *
                      (material_lookup[i+1]->specific_heat(temperature,pressure) - background_cp);
    }
```
But that would only remove one duplicated line of assignment and trade it for an additional line of if-statement. Do you think this is easier to understand? I feel my version version is easier, because the simplest case (`material_lookup.size() == 1`) only has a single instruction.

I could also move the additional if-statement of the alternative above into an additional exit condition in the for-loop, but that would make the logic even more obscure.